controller_manager/include/controller_manager/controller_manager.hpp

```diff
@@ -69,44 +84,211 @@ class ControllerManager : public rclcpp::Node
-  std::vector<std::shared_ptr<controller_interface::ControllerInterface>> loaded_controllers_;
+
+  /**
+   * @brief The RTControllerListWrapper class wraps a  double-buffered list of controllers
+   * to avoid needing to lock the real-time thread when switching controllers in
+   * the non-real-time thread.
+   *
+   * There's always an "updated" list and an "outdated" one
+   * There's always an "used by rt" list and an "unused by rt" list
+   *
+   * The updated state changes on the switch_updated_list()
+   * The rt usage state changes on the update_and_get_used_by_rt_list()
+   */
+  class RTControllerListWrapper
+  {
+public:
+    /**
+     * @brief update_and_get_used_by_rt_list Makes the "updated" list the "used by rt" list
+     * @warning Should only be called by the RT thread, no one should modify the
+     * updated list while it's being used
+     * @return reference to the updated list
+     */
+    std::vector<ControllerSpec> & update_and_get_used_by_rt_list();
+
+    /**
+     * @brief get_unused_list Waits until the "outdated" and "unused by rt"
+     *  lists match and returns a reference to it
+     * This referenced list can be modified safely until switch_updated_controller_list()
+     * is called, at this point the RT thread may start using it at any time
+     * @param guard Guard needed to make sure the caller is the only one accessing the unused by rt list
+     */
+    std::vector<ControllerSpec> & get_unused_list(
+      const std::lock_guard<std::recursive_mutex> & guard);
+
+    /**
+     * @brief get_updated_list Returns a const reference to the most updated list,
+     * @warning May or may not being used by the realtime thread, read-only reference for safety
+     * @param guard Guard needed to make sure the caller is the only one accessing the unused by rt list
+     */
+    const std::vector<ControllerSpec> & get_updated_list(
+      const std::lock_guard<std::recursive_mutex> & guard) const;
+
+    /**
+     * @brief switch_updated_list Switches the "updated" and "outdated" lists, and waits
+     *  until the RT thread is using the new "updated" list.
+     * @param guard Guard needed to make sure the caller is the only one accessing the unused by rt list
+     */
+    void switch_updated_list(const std::lock_guard<std::recursive_mutex> & guard);
+
+
+    // Mutex protecting the controllers list
+    // must be acquired before using any list other than the "used by rt"
+    mutable std::recursive_mutex controllers_lock_;
+
+private:
+    /**
+     * @brief get_other_list get the list not pointed by index
+     */
+    int get_other_list(int index) const;
+
+    void wait_until_rt_not_using(int index) const;
+
+    std::vector<ControllerSpec> controllers_lists_[2];
+    /// The index of the controller list with the most updated information
+    int updated_controllers_index_ = 0;
+    /// The index of the controllers list being used in the real-time thread.
+    int used_by_realtime_controllers_index_ = -1;
```

Comment:
Yeah, we can use two iterators instead of two ints, but it's always two variables to indicate:

- Which list is being used by RT thread (to know when a list is unused and delete it's contents from a non-RT thread)
- Which list should the RT thread use next

I may be missing something, but I believe we always need this "acknowledge" from the RT thread, because the non-RT thread is waiting for it to use the new list and cleanup the old one, and that requires two variables, or at least a more expressive variable.

But again, coming to my point for not rewriting this from scratch... This works, and has worked for the last decade, and is efficient. We have a lot of pending work on other fronts. Yes it's ugly and there's probably better solutions to it, but I feel like we're investing a lot of time for some code that will be forgotten like the old version was, because it works.

At least I suggest we move it to an issue and look at it again after this PR is done.

Reply:
> At least I suggest we move it to an issue and look at it again after this PR is done.

ok